test/blackbox/common/DDSBlackboxTestsDataReader.cpp

```diff
@@ -535,6 +535,306 @@ TEST(DDSDataReader, datareader_qos_use_topic_qos)
+ *
+ * get_matched_publications() must return RETCODE_OK
+ * with an empty list if no DataWriters are matched.
+ */
+TEST(DDSDataReader, datareader_get_matched_publications_ok_empty_list)
+{
+    PubSubReader<HelloWorldPubSubType> reader(TEST_TOPIC_NAME);
+    PubSubWriter<HelloWorldPubSubType> writer_1(TEST_TOPIC_NAME);
+    PubSubWriter<HelloWorldPubSubType> writer_2(TEST_TOPIC_NAME);
+
+    std::vector<InstanceHandle_t> pub_handles;
+
+    reader.reliability(RELIABLE_RELIABILITY_QOS)
+            .init();
+
+    writer_1.reliability(BEST_EFFORT_RELIABILITY_QOS)
+            .init();
+
+    writer_2.ownership_strength(10)
+            .init();
+
+    ASSERT_TRUE(reader.isInitialized());
+    ASSERT_TRUE(writer_1.isInitialized());
+    ASSERT_TRUE(writer_2.isInitialized());
+
+    // Reader should not be matched with any writer
+    reader.wait_discovery(std::chrono::seconds(2), 2);
+    ASSERT_FALSE(reader.is_matched());
+
+    auto& native_reader = reader.get_native_reader();
+    ReturnCode_t ret = native_reader.get_matched_publications(pub_handles);
+
+    ASSERT_EQ(ret, ReturnCode_t::RETCODE_OK);
+    ASSERT_EQ(pub_handles.size(), 0u);
+}
+
+/**
+ * Refers to DDS-DR-API-GMP-02 from the test plan.
+ *
+ * get_matched_publications() must provide the correct list of matched publication handles.
+ * Parameterize the test for different transports.
+ */
+TEST_P(DDSDataReader, datareader_get_matched_publications_correctly_behaves)
+{
+    const size_t num_writers = 5;
+
+    PubSubReader<HelloWorldPubSubType> reader(TEST_TOPIC_NAME);
+    std::vector<std::unique_ptr<PubSubWriter<HelloWorldPubSubType>>> writers;
+    std::vector<InstanceHandle_t> expected_pub_handles;
+    std::vector<InstanceHandle_t> pub_handles;
+
+    writers.reserve(num_writers);
+    pub_handles.reserve(num_writers);
+
+    reader.reliability(RELIABLE_RELIABILITY_QOS)
+            .init();
+
+    ASSERT_TRUE(reader.isInitialized());
+
+    for (size_t i = 0; i < num_writers; ++i)
+    {
+        writers.emplace_back(new PubSubWriter<HelloWorldPubSubType>(TEST_TOPIC_NAME));
+        writers.back()->init();
+        ASSERT_TRUE(writers.back()->isInitialized());
+        expected_pub_handles.emplace_back(writers.back()->get_native_writer().get_instance_handle());
+    }
+
+    // Wait for discovery
+    reader.wait_discovery(std::chrono::seconds::zero(), num_writers);
+    ASSERT_EQ(reader.get_matched(), num_writers);
+
+    auto& native_reader = reader.get_native_reader();
+    ReturnCode_t ret = native_reader.get_matched_publications(pub_handles);
+
+    // Check that the list of matched publication handles is correct
+    ASSERT_EQ(ret, ReturnCode_t::RETCODE_OK);
+    ASSERT_EQ(pub_handles.size(), num_writers);
+    ASSERT_TRUE(std::is_permutation(pub_handles.begin(), pub_handles.end(), expected_pub_handles.begin()));
+
+    // Remove two writers and check that the list of matched publication handles is updated
+    writers.pop_back();
```

Comment:
I think this doesn't destroy the writer. You have to do a `delete`.

Reply:
I think it is being destroyed as it is a `unique_ptr`, so when the pointer is removed from the collection it will be automatically be destroyed and will call the `~PubSUbWriter<>`